service/matching/physical_task_queue_manager.go

```diff
@@ -540,6 +527,79 @@ func (c *physicalTaskQueueManagerImpl) TrySyncMatch(ctx context.Context, task *i
 	return c.matcher.Offer(childCtx, task)
 }
 
+func (c *physicalTaskQueueManagerImpl) ensureRegisteredInDeployment(
+	ctx context.Context,
+	namespaceEntry *namespace.Namespace,
+	pollMetadata *pollMetadata,
+) error {
+	if !pollMetadata.workerVersionCapabilities.UseVersioning {
+		return nil
+	}
+	if !c.partitionMgr.engine.config.EnableDeployments(namespaceEntry.Name().String()) {
+		return nil
+	}
+
+	// lock so that only one poll does the update and the rest wait for it
+	c.deploymentLock.Lock()
+	defer c.deploymentLock.Lock()
+
```

Comment:
I agree about mixing locks and atomics, that's why there are no more atomics.

if one poll comes in and starts the workflow/update, and another 999 come in in the next half second, there should only be one outstanding call. otherwise we'll slam frontend+history.

actually we need even more backoff here on error (comment below), but that can wait.

Reply:
oops, thought you had decided to keep `c.deploymentRegistered` an atomic. The git diff confused me, all good. I like this approach of using one and not both.

thanks for the reasoning but then we should document somewhere that the poll latency (using versioning) might increase given we are waiting on userdata propogating and these calls to frontend/matching

Reply:
only on the first few polls from a new deployment. it's a potential failure mode but it shouldn't affect latency in the steady state.